contracts/main/CurveCryptoMathOptimized3.vy

```diff
@@ -552,220 +550,55 @@ def get_p(
+    x: int256 = convert(_xp[0], int256)
+    y: int256 = convert(_xp[1], int256)
+    z: int256 = convert(_xp[2], int256)
 
-    # (10**18 + gamma)*(-10**18 + gamma*(-2*10**18 + (-10**18 + 10**18*A/10000)*gamma/10**18)/10**18)/10**18
-    # this entire expression can be unsafe:
-    s1: int256 = unsafe_div(
-        unsafe_mul(
-            unsafe_add(10**18, gamma),
-            unsafe_add(
-                -10**18,
-                unsafe_div(
-                    unsafe_mul(
-                        gamma,
-                        unsafe_add(
-                            -2 * 10**18,
-                            unsafe_div(
-                                unsafe_mul(
-                                    unsafe_add(
-                                        -10**18,
-                                        unsafe_div(
-                                            unsafe_mul(10**18, ANN),
-                                            convert(A_MULTIPLIER, int256),
-                                        ),
-                                    ),
-                                    gamma,
-                                ),
-                                10**18,
-                            ),
-                        ),
-                    ),
-                    10**18,
-                ),
-            ),
-        ),
-        10**18,
-    )
-
-    # 81*(10**18 + gamma*(2*10**18 + gamma + 10**18*9*A/27/10000*gamma/10**18)/10**18)*x1/D*x2/D*x3/D
-    # Multiplication by x1, x2 and x3 needs to be safe:
-    s2: int256 = unsafe_div(
-        unsafe_mul(
-            gamma,
-            unsafe_add(
-                unsafe_add(2 * 10**18, gamma),
-                unsafe_div(
-                    unsafe_mul(
-                        unsafe_div(
-                            unsafe_div(
-                                unsafe_mul(10**18 * 9, ANN), 27
-                            ),
-                            convert(A_MULTIPLIER, int256),
-                        ),
-                        gamma,
-                    ),
-                    10**18,
-                ),
-            )
-        ),
-        10**18,
-    )
-    s2 = unsafe_div(
-        unsafe_div(
-            unsafe_div(
-                unsafe_mul(81, (10**18 + s2)) * x1, D
-            ) * x2,
-            D,
-        ) * x3,
-        D
-    )
+    S: int256 = unsafe_add(unsafe_add(x, y), z)
 
-    # 2187*(10**18 + gamma)*x1/D*x1/D*x2/D*x2/D*x3/D*x3/D
-    # Same as before: multiplication by x1, x2 or x3 should be safe
-    s3: int256 = unsafe_mul(2187, unsafe_add(10**18, gamma))
-    s3 = unsafe_div(
-        unsafe_div(
-            unsafe_div(s3 * x1, D) * x2, D
-        ) * x3,
-        D
-    )
-    s3 = unsafe_div(
-        unsafe_div(
-            unsafe_div(s3 * x1, D) * x2, D
-        ) * x3,
-        D
-    )
+    # constant which is in 10**36 precision:
+    NN_A_gamma2: int256 = unsafe_mul(27 * ANN, unsafe_mul(gamma, gamma))
```

Comment:
this can be optimized if you declare `NN_A_gamma2: int256 = unsafe_div(unsafe_mul(ANN, unsafe_mul(gamma, gamma)), 10000)`